netbox/utilities/utils.py

```diff
@@ -377,20 +377,32 @@ def prepare_cloned_fields(instance):
     return QueryDict(urlencode(params), mutable=True)
 
 
-def shallow_compare_dict(source_dict, destination_dict, exclude=tuple()):
-    """
-    Return a new dictionary of the different keys. The values of `destination_dict` are returned. Only the equality of
-    the first layer of keys/values is checked. `exclude` is a list or tuple of keys to be ignored.
-    """
-    difference = {}
-
-    for key, value in destination_dict.items():
-        if key in exclude:
-            continue
-        if source_dict.get(key) != value:
-            difference[key] = value
-
-    return difference
+def deep_compare_dict(old, new):
+    """
+    Return a tuple of two dictionaries `(removed_diffs, added_diffs)` in a format
+    that is compatible with the requirements of `ObjectChangeView`.
+    """
+    added_diffs = {}
+    removed_diffs = {}
+
+    for k in old:
+        old_data = old[k]
+        new_data = new[k]
+        # HACK: Sometimes, there are issues with floating point fields
+        # where the type of old_data is `float` and new_data is a `str`.
+        # So we serialize it to a string and compare it.
```

Comment:
I feel this needs much more consideration. A change in data type is a change in data; we should not be hand-waving this. The current implementation does not force string casting when evaluating data, so neither should any improvements.